### ISSUE TITLE
Slow queries (Next.js + Prisma + Postgres)

### ISSUE THREAD

**Issue by albingroen** [prisma/prisma#7009]
### Bug description

I have built an app where you can house forums. This application is built using Next.js, Prisma, and Postgres. Everything runs super smooth and fast in the local development environment.

- [x]  When I deployed everything live I ran into a couple of issues. I resolved the database connection issues I was having using connection pools for the Postgres database in Digitalocean.

- [x]  I resolved some performance issues I was having by making sure that all services (client/serverless functions/database) are located in the same geographical region (Germany/Belgium)

## However

I am still having some pretty annoying performance issues. It's on the border of not being a particularly good user experience. I have run some tests using k6 on a couple of different test routes, to demonstrate the issues I'm having.

![CleanShot 2021-05-10 at 11 22 09](https://user-images.githubusercontent.com/19674362/117637315-fab51100-b181-11eb-8a5d-15ee89dcc1a0.png)

### Get forums

```javascript
export default async (req, res) => {
  const forums = await prisma.forum.findMany({
    include: {
      users: true,
    },
  });

  res.json(forums);
};
```

### Get forums without users

```javascript
export default async (req, res) => {
  const forums = await prisma.forum.findMany();

  res.json(forums);
};
```

### Get without query
```javascript
export default async (req, res) => {
  res.status(200).end()
};
```

### How to reproduce

1. `git clone git@github.com:albingroen/formulate.git`
2. `cd app && yarn`
3. [Install k6](https://k6.io/docs/getting-started/installation/)
4. `env API_URL="https://formulate-theta.vercel.app/api" k6 run loadTest.js`

### Expected behavior

I expect the requests (*Get forums* and *Get forums without users*) to be quite a bit faster. The tables include less than 10 rows each, so that shouldn't be the reason either.

### Prisma information

```
datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

generator client {
  provider = "prisma-client-js"
}

model PostReaction {
  id    Int    @id @default(autoincrement())
  value String @db.VarChar(255)

  post   Post @relation(fields: [postId], references: [id])
  postId Int

  user   User @relation(fields: [userId], references: [id])
  userId Int
}

model Post {
  id        Int      @id @default(autoincrement())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
  title     String   @db.VarChar(255)
  content   String?
  published Boolean  @default(false)
  posts     Post[]   @relation("PostToPost")
  user      User     @relation(fields: [userId], references: [id])
  userId    Int

  forum   Forum @relation(fields: [forumId], references: [id])
  forumId Int
  Post    Post? @relation("PostToPost", fields: [postId], references: [id])
  postId  Int?

  reactions PostReaction[]

  @@map(name: "posts")
}

enum ForumUserRole {
  USER
  ADMIN
}

model ForumUser {
  id        Int           @id @default(autoincrement())
  createdAt DateTime      @default(now())
  updatedAt DateTime      @updatedAt
  user      User          @relation(fields: [userId], references: [id])
  role      ForumUserRole
  userId    Int

  forum   Forum? @relation(fields: [forumId], references: [id])
  forumId Int?

  @@map(name: "forum_users")
}

model Forum {
  id          Int         @id @default(autoincrement())
  createdAt   DateTime    @default(now())
  updatedAt   DateTime    @updatedAt
  title       String      @db.VarChar(255)
  description String?
  posts       Post[]
  users       ForumUser[]
  logotype    String?

  user   User? @relation(fields: [userId], references: [id])
  userId Int?

  Invitation Invitation[]
  @@map(name: "forums")
}

model Account {
  id                 Int       @id @default(autoincrement())
  compoundId         String    @unique @map(name: "compound_id")
  userId             Int       @map(name: "user_id")
  providerType       String    @map(name: "provider_type")
  providerId         String    @map(name: "provider_id")
  providerAccountId  String    @map(name: "provider_account_id")
  refreshToken       String?   @map(name: "refresh_token")
  accessToken        String?   @map(name: "access_token")
  accessTokenExpires DateTime? @map(name: "access_token_expires")
  createdAt          DateTime  @default(now()) @map(name: "created_at")
  updatedAt          DateTime  @default(now()) @map(name: "updated_at")


  @@index([providerAccountId], name: "providerAccountId")
  @@index([providerId], name: "providerId")
  @@index([userId], name: "userId")
  @@map(name: "accounts")
}

model Session {
  id           Int      @id @default(autoincrement())
  userId       Int      @map(name: "user_id")
  expires      DateTime
  sessionToken String   @unique @map(name: "session_token")
  accessToken  String   @unique @map(name: "access_token")
  createdAt    DateTime @default(now()) @map(name: "created_at")
  updatedAt    DateTime @default(now()) @map(name: "updated_at")

  @@map(name: "sessions")
}

model User {
  id            Int       @id @default(autoincrement())
  name          String?
  email         String?   @unique
  emailVerified DateTime? @map(name: "email_verified")
  image         String?
  createdAt     DateTime  @default(now()) @map(name: "created_at")
  updatedAt     DateTime  @default(now()) @map(name: "updated_at")

  postReactions PostReaction[]
  forumUsers ForumUser[]
  forums     Forum[]
  posts      Post[]

  @@map(name: "users")
}

model VerificationRequest {
  id         Int      @id @default(autoincrement())
  identifier String
  token      String   @unique
  expires    DateTime
  createdAt  DateTime @default(now()) @map(name: "created_at")
  updatedAt  DateTime @default(now()) @map(name: "updated_at")

  @@map(name: "verification_requests")
}

enum InvitationStatus {
  PENDING
  DECLINED
  ACCEPTED
}

model Invitation {
  id     Int              @id @default(autoincrement())
  status InvitationStatus @default(PENDING)

  forum   Forum @relation(fields: [forumId], references: [id])
  forumId Int
}

```

### Environment & setup

- OS: MacOS
- Database: PostgreSQL
- Node.js version: 12.8.2
- Prisma version:
```
prisma               : 2.22.1
@prisma/client       : 2.22.1
Current platform     : darwin
Query Engine         : query-engine 60cc71d884972ab4e897f0277c4b84383dddaf6c (at node_modules/@prisma/engines/query-engine-darwin)
Migration Engine     : migration-engine-cli 60cc71d884972ab4e897f0277c4b84383dddaf6c (at node_modules/@prisma/engines/migration-engine-darwin)
Introspection Engine : introspection-core 60cc71d884972ab4e897f0277c4b84383dddaf6c (at node_modules/@prisma/engines/introspection-engine-darwin)
Format Binary        : prisma-fmt 60cc71d884972ab4e897f0277c4b84383dddaf6c (at node_modules/@prisma/engines/prisma-fmt-darwin)
Default Engines Hash : 60cc71d884972ab4e897f0277c4b84383dddaf6c
Studio               : 0.379.0
```

**Comment by pantharshit00** [comment]
Hello @albingroen 

I wasn't able to reproduce this. Here is results of my production deploy:
![image](https://user-images.githubusercontent.com/22195362/117677683-4db3b600-b1cc-11eb-9cec-8cd519644b3c.png)

The average is around 100ms which is acceptable


This is using a digitalocean database in sfo3 region and a vercel deployment to sfo1 region. Latency is under 10ms for most of the cases. This was tested from a machine in sfo.  Here is my deployment: https://issue-7009-drigger.vercel.app/api/test


Can you please double check the database region and serverless region once? Also, can you please check the lambda execution time. Here is what I am getting there:
![image](https://user-images.githubusercontent.com/22195362/117677947-8ce20700-b1cc-11eb-8ff7-976d2d1dd77d.png)
![image](https://user-images.githubusercontent.com/22195362/117677964-92d7e800-b1cc-11eb-8816-497bd25eb44f.png)


More things to double check:
- Did you use the connection string for pgbouncer in your app? Digitalocean creates a new connection string when you create a connection pool. You need to use that in order to use pgbouncer. 
- Did you add `&pgbouncer=true` to your connection url in order to enable pgbouncer mode in Prisma?

**Comment by pantharshit00** [comment]
Yeah, if you run load test from far away, network latency will be there. 


Can you please check the lamba execution time like I have done in the first comment. That's the most important metric here.

**Comment by albingroen** [comment]
What do you use to test that?

**Comment by albingroen** [comment]
Anyways, here's a local run with k6 against `https://issue-7009-drigger.vercel.app/api/test`

<img width="1442" alt="CleanShot 2021-05-10 at 18 43 06@2x" src="https://user-images.githubusercontent.com/19674362/117694613-a16bd280-b1bf-11eb-948e-7ce1ecc67b9b.png">

**Comment by pantharshit00** [comment]
I used vercel dashboard for that. Goto logs section of your deployment and click show details. Then you can trigger requests to see the execution time. 


If that is off I will try a deployment in EU tomorrow now for a reproduction.

**Comment by albingroen** [comment]
Ah okay got you!

<img width="549" alt="CleanShot 2021-05-10 at 19 00 37@2x" src="https://user-images.githubusercontent.com/19674362/117696893-1dffb080-b1c2-11eb-8952-008996bfd62c.png">

<img width="494" alt="CleanShot 2021-05-10 at 19 02 37@2x" src="https://user-images.githubusercontent.com/19674362/117697036-4e474f00-b1c2-11eb-99f5-c5acf85dd4e2.png">

**Comment by albingroen** [comment]
That's really interesting. `/api/test` basically does the same thing as `/api/forums` and it's a HUGE difference. This one was triggered from me loading forums in the actual application.

<img width="1116" alt="CleanShot 2021-05-10 at 19 03 55@2x" src="https://user-images.githubusercontent.com/19674362/117697262-91a1bd80-b1c2-11eb-863e-9e2711a43f41.png">

**Comment by sorenbs** [comment]
At this point it seems that you have ruled out the serverless environment and connection pooling as likely issues. 

The next step is to look at the actual queries and verify that your database is able to execute them quickly. Anything above 10 ms is probably too slow. 

Q1: have you tried running this sql query directly against the database using the database cli or a GUI tool? You can see her how to make Prisma print the SQL queries it generates: https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/logging

Q2: do you have large amounts of data? If you have a million records or more, it’s important to configure indexes correctly to support your queries. 

Lastly, I’ll just mention that it is really important for Prisma to be deployed as close to the database as possible. Ideally you should aim for 10 ms network latency or less. This is because Prisma often issues multiple consecutive requests, so you will incur a latency cost that is a multiple of the network latency between Prisma and the database. 

Hope this helps.

**Comment by albingroen** [comment]
@sorenbs I have done some testing on this now by running the queries directly against the database. I can confirm that more or less every call I make takes around 50ms. I have basically no data at all. Literally just 1 row in the forums table, and a handful of rows in the user table.

Right now have the database positioned in Frankfurt and Prisma (Next.js serverless functions) in Brussels. That was as close as I could get them.

**Comment by sorenbs** [comment]
Thank you!  For reference, the latency between Frankfurt and Brussels on the public Internet is about 20 ms assuming you already have an established connection: https://wondernetwork.com/pings/Frankfurt/Brussels

**Comment by janpio** [comment]
Is there anything else we can do to help here @albingroen?

We know Prisma is not doing a great job there helping you to figure out these things yourself, and in theory we could do that in the future as we sit "between" your application and database. We will try to start with some documentation that basically includes the advice you got above - put the database and application servers/functions really close to each other, make sure you attribute for the normal latency from your location and so on.

**Comment by albingroen** [comment]
I did some more measurements, and I think I found the trouble maker to be next-auth.

```
--- With exact same code on both routes

/api/test:
138.82 ms

/api/forums:
136.48 ms

--- With exact same code on both routes and with auth

/api/test
412.83 ms

/api/forums
433.92 ms
```

The first set is also kind of slow, but that's because of the double include of `ForumToUser` and `User`. But I think this can kind of conclude that I

1. Should perhaps use another authentication solution
2. Maybe opt for making more shallow includes

What do you guys think?

**Comment by janpio** [comment]
Yep, that does indeed not look good. Are you using the [Prisma adapter for NextAuth](https://next-auth.js.org/schemas/adapters#prisma-adapter)?

Is your code of that simple enough to add to your reproduction so we can play with it again? Maybe even exactly the two cases - with and without.

**Comment by albingroen** [comment]
Yes, I'm using the Prisma adapter. I can say that moving from using session-based authentication to JWTs helped a lot

```
--- With session-based auth

/api/forums:
433.92 ms

--- With JWT based auth

/api/forums:
131.77 ms
```

I've removed the test routes now, but if you want to I could set up another project with the same schemas and test routes that you can play with

**Comment by janpio** [comment]
Yes that would be wonderful. We are debugging next-auth and the Prisma adapter now really, but the authors are friends of the company of course so this would still be useful.

**Comment by albingroen** [comment]
Here, I set up a new (more minimal) repo where you can test: https://github.com/albingroen/prisma-next-auth-perf-test

**Comment by albingroen** [comment]
If you run it you'll find 3 buttons on the index page that you can use to test too

**Comment by sorenbs** [comment]
Thank you for thoroughly investigating this issue @kripod!

It seems like the NextAuth maintainer is open to potentially change the implementation, so hopefully this will get better soon. Glad you found a different approach that works for you!

**Comment by balazsorban44** [comment]
NextAuth maintainer here. I AM open for a different implementation, and I did tinker a bit, and created a temporary version at `3.22.0-canary.2`

```js
import {getServerSession} from "next-auth"
import {options} from "pages/api/auth/[...nextauth]"
export default Page(){
  return null
}
export async function getServerSideProps(context) {
  return {
    props: {
      session: await getServerSession(context, options),
    },
  }
}
```

And in your `[...nextauth].js` you have to extract the options into its own const, so you can share it in the page file. This configuration could ultimately be extracted into something like `next-auth.config.js` or similar I guess, but this is a Work In Progress for now.

Very interested if someone could check out if it helps. the `getServerSession` does not do `fetch`
